Feature/clover reconstruct

maddyscientist opened the pull request:
Feature-add branch
* Add clover-reconstruct CMake option `QUDA_CLOVER_RECONSTRUCT` to QUDA (enabled by default) which compresses the clover field when using the native field order.  This is only supported when `QUDA_CLOVER_DYNAMIC_INVERSE=ON`
* This reduces the memory footprint and improves performance for all kernels that utilize clover fields
* Clean up of handling when dynamic-inverse is enabled to remove hacks in the interface
* Dynamic-clover now utilizes double precision for the Cholesky for double and single precision fields, and single precision for half and quarter precision fields.  This improves solver stability.
* Unified `cpuCloverField` and `cudaCloverField`: we now only have `CloverField`
* Add support for `FLOAT8` ordering for half and quarter precision clover fields
* Utilize global norm scale for fixed-point clover field instead of per-site normalization
* Fixes to tunable_reduction.h with respect to correct setting of the `aux` key
* Remove legacy `writeAuxString` method

maddyscientist committed:
Add new Cmake option QUDA_CLOVER_RECONSTRUCT option to enabling compressed clover storage.  Renamed prior option QUDA_DYNAMIC_CLOVER to QUDA_CLOVER_DYNAMIC_CLOVER for better ccmake catergorization.  These options are now both enabled by default.

maddyscientist committed:
Fix pass at adding compressed clover storage access in accessor.  Only site-level support enabled for now (no fine grain support)

maddyscientist committed:
Unification of cpuCloverField and cudaCloverField, we now have only CloverField.  Move constexpr clover helper functions into the clover namespace.  When clover reconstruction is enabled, we can now correctly compute the inverse for return

maddyscientist committed:
Fix bug in CloverField::copy

maddyscientist committed:
Improved clover reconstruction

maddyscientist committed:
Cleanup the various hacks in the clover field handling, removing the aliasing fields.  For copy back to host, the reordering can now done on the device.

maddyscientist committed:
Fix clover application for direct term when dynamic clover is enabled

maddyscientist committed:
Merge branch 'feature/generic_kernel' of github.com:lattice/quda into feature/clover-reconstruct

maddyscientist committed:
Merge branch 'feature/generic_kernel' of github.com:lattice/quda into feature/clover-reconstruct

maddyscientist committed:
Add clover reconstruct support to fine-grain reader

maddyscientist committed:
Fix local memory for fine-grain accessors: chirality needs to be pulled out of spin to allow for loop unrolling.  Fix bytes computation for computeCoarse

maddyscientist committed:
Fine-grain clover accessor now does reconstuction

maddyscientist committed:
Only read and write clover compressed components

maddyscientist committed:
We now only store 28 numbers per chiral block for Nc=3 clover when reconstruct is enabled, reducing the memory footprint

maddyscientist committed:
Add missing unroll statement to clover invert

maddyscientist committed:
Add float-8 ordering support for fixed-precision clover

maddyscientist committed:
Fix for clover creation in coarse_op.cu

maddyscientist committed:
Further cleanup of historical hacks.  We now always allocate the direct clover field, with optional inverse allocation depending if doing dynamic inversion or if not required (e.g., not doing preconditioning)

maddyscientist committed:
Remove norm field from clover and instead use a global norm (different for direct and inverse).  Clover computation and inversion is now done in the callee precision, before copy down to desired precision.  General cleanup and removal of some hacks

maddyscientist committed:
Some cleanup

maddyscientist committed:
Fix bug in dslash_clover_helper.cuh: to prevent false error warning, cloverInv accessor must not use inverse field if dynamic inversion is enabled

maddyscientist committed:
Reconstruct clover now stores the clover field without the identity, this improves the precision of the fixed point formats.  When using clover reconstruct, we now need to set the diagonal coefficient.  When copying from a non-reconstruct field, this is done by running a micro kernel to compute this prior to the copy

maddyscientist committed:
Created a new trait clover::cholesky_t that sets the precision to use for the cholesky precision.  At the moment this is always double precision regardless of the kernel precision.  Add new solve function that wraps both the backward and forward substitution, for a simpler external interface

maddyscientist committed:
Fix copy routines in loadSloppyCloverQuda which led to erroneous behaviour in asymmetric preconditioning with mixed precision solvers

maddyscientist committed:
Fix bug in setting clover refinement precision in tests

maddyscientist committed:
Updated precision traits for Cholesky: half and quarter precision storage dslashes now use single precision

maddyscientist committed:
Remove writeAuxString function, and at the same time fix bug with tuneKey setting in cloverInvert

maddyscientist committed:
When copying from a compressed clover field to another, copy the raw elements instead of reconstructing and re-compressing: this reduces precision loss for lower precision

maddyscientist committed:
Fix bug in TunableReduction where aux string was not being fully populated.

maddyscientist committed:
Apply clang format

maddyscientist committed:
Merge branch 'develop' into feature/clover-reconstruct

maddyscientist committed:
Fix clover prefetching for when dynamic_clover is enabled

maddyscientist committed:
Apply clang format

weinbe2 commented:
> Dynamic-clover now utilizes double precision for the Cholesky for double and single precision fields

While I understand the sentiment of stability, I'm not sure I like the "silent fp64" in the single precision operator.

maddyscientist committed:
Fix clang compilation warning

maddyscientist committed:
Fix unused typedef warning on clang

weinbe2 commented:
> > Dynamic-clover now utilizes double precision for the Cholesky for double and single precision fields
> 
> While I understand the sentiment of stability, I'm not sure I like the "silent fp64" in the single precision operator.

Conversation resolved offline. While I still don't necessary like the "silent fp64" in a single precision operator, there's still an option to compile w/out dynamic clover -> the inverse clover gets precomputed and directly applied in fp32, giving a way out if avoiding fp64 is truly desired.

weinbe2 commented:
I've gone through a full cursory review and I'm very impressed with this PR. I'm about to dig into some hands-on testing

one request I have in the interim is for some documentation on these new features on the wiki; some type of "clover fermion best practices". This could go under `QUDA Solver Acceleration Strategies` (similar to documentation for advanced dwf features, etc) or `Using QUDA in your LQCD applications`, or wherever else is deemed fit.

cpviolator commented on include/clover_field.h:
```diff
@@ -32,12 +79,10 @@ namespace quda {
   struct CloverFieldParam : public LatticeFieldParam {
-    bool direct; // whether to create the direct clover 
-    bool inverse; // whether to create the inverse clover
+    bool reconstruct; /** Whether to create a compressed that requires reconstruction */
```
"a compressed clover field that"?

maddyscientist replied:
84359ab

cpviolator commented on include/clover_field_order.h:
```diff
@@ -519,49 +558,60 @@ namespace quda {
+        typedef typename AllocType<huge_alloc>::type AllocInt;
+        typedef float norm_type;
+        static constexpr int Ns = 4;
+        static constexpr int Nc = 3;
```
Can you place a static_assert here on 3?

maddyscientist replied:
84359ab

cpviolator commented on include/enum_quda_fortran.h:
```diff
@@ -237,9 +237,10 @@
-#define QUDA_PACKED_CLOVER_ORDER 5    // even-odd packed
-#define QUDA_QDPJIT_CLOVER_ORDER 6 // lexicographical order packed
-#define QUDA_BQCD_CLOVER_ORDER 7 // BQCD order which is a packed super-diagonal form
+#define QUDA_FLOAT8_CLOVER_ORDER 8   // even-odd float4 ordering
```
"// even-odd float8 ordering"

maddyscientist replied:
84359ab

cpviolator commented on include/enum_quda.h:
```diff
@@ -254,6 +254,7 @@ typedef enum QudaCloverFieldOrder_s {
   QUDA_FLOAT_CLOVER_ORDER = 1,  // even-odd float ordering
   QUDA_FLOAT2_CLOVER_ORDER = 2, // even-odd float2 ordering
   QUDA_FLOAT4_CLOVER_ORDER = 4, // even-odd float4 ordering
+  QUDA_FLOAT8_CLOVER_ORDER = 8, // even-odd float4 ordering
```
"// even-odd float8 ordering"

maddyscientist replied:
84359ab

cpviolator commented on include/clover_field_order.h:
```diff
@@ -881,20 +1026,36 @@ namespace quda {
+#ifdef FLOAT8
+#define N8 8
+#else
+#define N8 4
```
Can this be changed to something like

```C++
#ifdef FLOAT8
 #define FLOATN 8
 #else
 #define FLOATN 4
```
with the appropriate substitutions to `N8` below?

maddyscientist replied:
84359ab (and I applied the same cleanup to gauge_field_order.h and color_spinor_field_order.h)

cpviolator reviewed (changes requested):
All looks good to me. There some instances of Nc=3 specific code, but I can see most of them from the hardcoded values of 6, 30, 72, etc, so not too hard to upgrade the loops in the Nc agnostic branch.

maddyscientist committed:
Address @cpviolator's review comments

maddyscientist committed:
Guard the comm_broadcast to ensure it's not called while tuning - which otherwise causes a mismatched of MPI broadcasts between process 0 and all others

maddyscientist commented:
Thanks @cpviolator for the review.  I've addressed all your comments in 84359ab.

Also fixed a nasty UB bug with MPI broadcasts getting mixed up in d71d9d0.  PR revalidated, both single and multi-GPU.

mathiaswagner commented on include/instantiate.h:
```diff
@@ -151,6 +151,31 @@ namespace quda
+     @brief This instantiate2 function is used to instantiate the
+     precisions, with double precision always enabled.  This is a
+     temporary addition until we fuse this with the original function
+     above when we enforce C++17
```
What is stopping us from just enforcing C++17 now?

maddyscientist replied:
Nothing stopping us.  I would probably prefer C++17 related cleanup in a separate PR in favor of getting this merged first.

weinbe2 replied:
> Nothing stopping us. I would probably prefer C++17 related cleanup in a separate PR in favor of getting this merged first.

+1 on making it separate

mathiaswagner commented on CMakeLists.txt:
```diff
@@ -474,6 +476,10 @@ if(QUDA_HETEROGENEOUS_ATOMIC AND QUDA_JITIFY)
   message(SEND_ERROR "QUDA_HETEROGENEOUS_ATOMIC=ON does not support JITIFY)")
 endif()
 
+if(QUDA_CLOVER_RECONSTRUCT AND NOT QUDA_CLOVER_DYNAMIC)
```
thoughts on replacing this if by using
```
cmake_dependent_option(QUDA_CLOVER_RECONSTRUCT "set to ON to enable compressed clover storage" OFF "QUDA_CLOVER_DYNAMIC" OFF)
```
?
```

maddyscientist replied:
No strong thoughts.  Personally
I don’t like options disappearing from `ccmake` when certain options are set, and I like to see them, even if they error when enabled / disabled.

mathiaswagner replied:
works for me. I was just annoyed that to disable dynamic I had to switch to advanced to disable reconstruct separately.

maddyscientist committed:
Small cleanup of clover::FloatNOrder::load and save functions

maddyscientist commented:
Noting that I've applied some cleanup to the `clover::FloatNOrder::load` and `clover::FloatNOrder::save` functions to reduce code duplication and improve readability.  As suggested above, I've also made a wiki page to discuss some of the clover fermion options: https://github.com/lattice/quda/wiki/Clover-Fermions.

maddyscientist committed:
Fix for jitify

weinbe2 commented:
Making sure I understand what I need to test the outer product of:
* FLOAT8 on/off
* (dynamic off/reconstruct off) ; (dynamic on/reconstruct off) ; (dynamic on/reconstruct on)
Which is 6 combinations total. Am I missing any combination?

Also---do these different combinations get appropriately distinguished in tunekeys?

maddyscientist commented:
@weinbe2 your combinations look good.

Regarding whether these combinations are distinguished with tunekeys: in keeping in general with other compile-time options as opposed to run-time options, there is no effort to break degeneracy

weinbe2 reviewed (approved):
For the above six combinations of compile flags, I ran:
* A full `ctest`
* A separate test of multigrid across precisions, preconditionings, clover/twisted-clover, and Nc

I did not test different partitions because the clover field is an inherently rank-local quantity.

Everything passed and looked good, so this gets my approve!

cpviolator reviewed (approved):
All good! Awesome PR @maddyscientist
